### ISSUE TITLE
raisesubs doesn't factor in osc scaling

### ISSUE THREAD

**Issue by Keith94** [Samillion/ModernZ#24]
The subs can appear in a different position depending on if you're using non-default scaling

1. Set `scalefullscreen` to `0.5`
2. Enter fullscreen mode
3. Move mouse so OSC appears while subtitles on screen
4. Exit fullscreen. As a result, the subtitles appear closer to the bottom edge of screen

Windowed (before):

![mpv_qfMeAVDTGm](https://github.com/user-attachments/assets/17e3d469-0a8f-4c2d-b1ad-9942f9f2e99a)

Toggle fullscreen on and off:

![mpv_2pwuogwYRo](https://github.com/user-attachments/assets/1e297e2d-ed8e-46e8-a141-495b41646672)

At the same time the subtitles appear really high when you first enter fullscreen, and then kind of "jump" back down to the correct position

![mpv_57wCaMJDTn](https://github.com/user-attachments/assets/c9864a1f-6926-44e1-92d2-770c4784d6ba)

![mpv_KjSeHTqDG8](https://github.com/user-attachments/assets/fc785fda-03ff-44d2-a4f9-290a7707705a)

**Comment by Keith94** [comment]
The problem obviously gets worse if you use more extreme scaling values like

scalewindowed=2.0
scalefullscreen=0.1

https://github.com/user-attachments/assets/e7e57217-207d-4fd2-98e1-6860a4543123

**Comment by Samillion** [comment]
Yeah, that is another remnant of the past with a logic that doesn't address scaling
```lua
local subpos = math.floor((osc_param.playresy - user_opts.raisesubamount)/osc_param.playresy*100)
```

I could improve the logic of scale with:
```lua
    if state.fullscreen then
        scale = user_opts.scalefullscreen
    else
        scale = user_opts.scalewindowed
    end

    local raise_factor = user_opts.raisesubamount

    -- adjustment logic: reduce raise for extreme scales
    if scale > 1 then
        raise_factor = raise_factor * (1 + (scale - 1) * 0.5)  -- Moderate increase when scale > 1
    elseif scale < 1 then
        raise_factor = raise_factor * scale  -- Scale down when scale < 1
    end

    -- Adjust the subtitle position based on scaled raise amount
    local subpos = math.floor((osc_param.playresy - raise_factor) / osc_param.playresy * 100)
```

In that scenario though, `raisesubamount` will be scaled with window/fullscreen scale, so if it's not 1, it'll decrease/increase, but be proportionate to the scale.

**Comment by Samillion** [comment]
Could you try this please: https://github.com/Samillion/ModernZ/blob/dev_raisesub_scale/modernz.lua

**Noteworthy:**
with a `scalewindowed=2.0` or `scalefullscreen=0.1`, `raisesubamount` is adjusted. So it is not a static value. It will display it to fit the scale using `raisesubamount` as a main calculation factor with the current window/fullscreen scale.

A scale of `1` makes everything perfect and exactly where you'd want them to be, this is more of a `~nearby` position adjustment, if scale is > or < `1`.

**Comment by Keith94** [comment]
Ok sweet, I'll try it when I'm back home. It's a lot to unpack so I could misunderstand the logic.

I suppose, even though I requested it earlier, raisesubamount wouldn't be needed if the subs got raised by a good amount. Last time I checked it didn't take into account the osc title being displayed/hidden, but I can't remember how it works at the moment.

**Comment by Samillion** [comment]
Awesome, thank you very much for testing. I'll use this time to try and improve the scale calculation logic.

**Comment by Samillion** [comment]
I improved the calculation logic so that the scale doesn't adjust `raisesubamount` by much: https://github.com/Samillion/ModernZ/commit/a4c85a05d64800ad67e44dc7e470e75cf0f5bfbb

https://github.com/Samillion/ModernZ/blob/dev_raisesub_scale/modernz.lua

**Comment by Keith94** [comment]
I think overall it works a lot better. Sometimes when you toggle out of fullscreen, the subtitles don't react to the OSC being visible, so they are hidden behind the OSC, even with raisesubs=yes.

![mpv_Q2vIyAN5EN](https://github.com/user-attachments/assets/b2defc1b-f3b7-4ba7-83d8-2688d3873c98)

If you use a smaller scale in fullscreen, entering fullscreen will have the subs pretty high, until the cursor autohides, then they're displayed at a better height after that point. Not sure if either of these are easily solvable!

![mpv_ms9M6bWTz0](https://github.com/user-attachments/assets/01843e30-bb79-4019-b7e1-439119fb8f08)

**Comment by Samillion** [comment]
> I think overall it works a lot better. Sometimes when you toggle out of fullscreen, the subtitles don't react to the OSC being visible, so they are hidden behind the OSC, even with raisesubs=yes.

Do they stay like that or just for a second?

The difference here is the change in `raisesubamount` calculation based on scale.

Consider the following:
- With a scale of 1 and raise amount 170. Change windowed/fullscreen results in the same values: OSC height, raise amount, scale.
- With a scale > or < 1, and raise amount 170. On change windowed/fullscreen, everything changes, and it adjusts to place subs accordingly.

I can't think of a way to make it more seamless without needing that one/two seconds to adapt, to be honest. I'm not saying there isn't one, I just can't think of one lol.

**Comment by Samillion** [comment]
Hmm, my answer might be confusing.

In simpler terms: It gets the new values after the fact, not during or while. So only after change from windowed => fullscreen, it calculates.

As the values are not pre-stored. mpv adjusts `osd/display-size` on action, this process can be seen with `--v` and even with `--no-config --v`

**Comment by Samillion** [comment]
Could you try this see if it's better when you can, please: https://github.com/Samillion/ModernZ/blob/dev_raisesub_scale/modernz.lua

**Comment by Keith94** [comment]
Nice, from my testing it's working much better now.

I only noticed a weird edge case, where raisesubs feature get a bit "confused" (?) if you toggle FS in the middle of an osc timeout.

Set `scalewindowed=2.2` and `scalefullscreen=0.5`. It happens with other values, too
enter fullscreen with f
as soon as osc begins to hide/timeout, exit fullscreen with f

result: the osc will re-appear if the mouse cursor is inside the window viewport, with subs behind it
it should look something like this and take a few tries to manifest

https://github.com/user-attachments/assets/7534b779-37f5-4aef-bfe6-1bdecc46426a

**Comment by Samillion** [comment]
Awesome! I'll merge this and work on the other issues.
